File: Bamilo/Controllers/SignUpViewController.m
//
//  SignUpViewController.m
//  Bamilo
//
//  Created by Ali Saeedifar on 2/8/17.
//  Copyright © 2017 Rocket Internet. All rights reserved.
//

#import "SignUpViewController.h"
#import "EmarsysPredictManager.h"
#import "PushWooshTracker.h"
#import "AuthenticationContainerViewController.h"
//Lagacy importing
#import "RICustomer.h"
#import "JAUtils.h"
#import "DataServiceProtocol.h"
#import "Bamilo-Swift.h"

#define cSignUpMethodEmail @"email"
#define cSignUpMethodGoogle @"sso-google"


@interface SignUpViewController() <PhoneVerificationViewControllerDelegate>
@property (nonatomic, weak) IBOutlet UITableView* tableView;
@property (nonatomic, strong) FormViewControl *formController;
@property (nonatomic, strong) FormItemModel *phoneField;
@property (nonatomic) BOOL isSubmissionButtonDisabled;
@end

@implementation SignUpViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.formController = [[FormViewControl alloc] init];
    self.formController.delegate = self;
    
    [self.tableView setContentInset: UIEdgeInsetsMake(28, 0, 0, 0)];
    self.formController.tableView = self.tableView;
    
    FormItemModel *email = [FormItemModel emailWithFieldName:@"customer[email]"];
    self.phoneField = [FormItemModel phoneWithFieldName:@"customer[phone]"];
    FormItemModel *melliCode = [FormItemModel nationalID:@"customer[national_id]"];
    FormItemModel *password = [FormItemModel passWordWithFieldName:@"customer[password]"];
    
    self.formController.submitTitle = STRING_SIGNUP;
    self.title = STRING_SIGNUP;
    self.formController.formModelList = [NSMutableArray arrayWithArray:@[ email, self.phoneField, melliCode, password ]];
    [self.formController setupTableView];
}

- (void)viewDidAppear:(BOOL)animated {
    self.isSubmissionButtonDisabled = NO;
    [self.formController registerForKeyboardNotifications];
}

- (void)viewDidDisappear:(BOOL)animated {
    [self.formController unregisterForKeyboardNotifications];
}

#pragma mark - formControlDelegate
- (void)formSubmitButtonTapped {
    if (![self.formController isFormValid] && !self.isSubmissionButtonDisabled) {
        [self.formController showAnyErrorInForm];
        return;
    }
    [self tryToSignupUser:self];
}

- (void)tryToSignupUser:(id<DataServiceProtocol>)target {
    [DataAggregator signupUser:target with:[self.formController getMutableDictionaryOfForm] completion:^(id data, NSError *error) {
        if(error == nil) {
            self.isSubmissionButtonDisabled = YES;
            //TODO: this is not a good approach which has been contracted with server
            // and it's better to be refactored in both platforms
            if ([data isKindOfClass:[ApiDataMessageList class]]) {
                ApiDataMessageList *messages = data;
                if (messages) {
                    if ([((ApiDataMessage *)messages.success.firstObject).reason isEqualToString:@"CUSTOMER_REGISTRATION_STEP_1_VALIDATED"]) {
                        NSLog(@"comes here");
                        [self.delegate wantsToShowTokenVerificatinWith:self phone: [self.phoneField getValue]];
                        return;
                    }
                }
            }
            //End of TODO
            [self bind:data forRequestId:0];
            //EVENT: SIGNUP / SUCCESS
            RICustomer *customer = [RICustomer getCurrentCustomer];
            [TrackerManager postEventWithSelector:[EventSelectors signupEventSelector] attributes:[EventAttributes signupWithMethod:cSignUpMethodEmail user:customer success:YES]];
            [EmarsysPredictManager setCustomer: customer];
            [PushWooshTracker setUserID:[RICustomer getCurrentCustomer].email];
            if (self.completion) {
                self.completion(AuthenticationStatusSignupFinished);
            } else if ([[MainTabBarViewController topViewController] isKindOfClass:[AuthenticationContainerViewController class]]) {
                [((UIViewController *)self.delegate).navigationController popViewControllerAnimated:YES];
            } else {
                [((UIViewController *)self.delegate).navigationController popWithStep:2 animated:YES];
            }
        } else {
            [TrackerManager postEventWithSelector:[EventSelectors signupEventSelector]
                                       attributes:[EventAttributes signupWithMethod:cSignUpMethodEmail user:nil success:NO]];
            //EVENT: SIGNUP / FAILURE
            if ([target isKindOfClass:[SignUpViewController class]]) {
                BaseViewController *baseViewController = (BaseViewController *)self.delegate;
                if(![baseViewController showNotificationBar:error isSuccess:NO]) {
                    BOOL errorHandled = NO;
                    NSArray<NSDictionary *> *errors = [error.userInfo objectForKey:kErrorMessages];
                    for(NSDictionary* errorField in errors) {
                        NSString *fieldNameParm = [errorField objectForKey:@"field"];
                        if ([fieldNameParm isKindOfClass:[NSString class]] && fieldNameParm.length > 0) {
                            NSString *fieldName = [NSString stringWithFormat:@"customer[%@]", fieldNameParm];
                            if ([self.formController showErrorMessageForField:fieldName errorMsg:errorField[kMessage]]) {
                                errorHandled = YES;
                            }
                        }
                    }
                    if (!errorHandled) {
                        [self showNotificationBarMessage:STRING_CONNECTION_SERVER_ERROR_MESSAGES isSuccess:NO];
                    }
                }
            } else if ([target isKindOfClass:[BaseViewController class]]) {
                if (![Utility handleErrorMessagesWithError:error viewController:(BaseViewController *)target]) {
                    [(BaseViewController *)target showNotificationBarMessage:STRING_CONNECTION_SERVER_ERROR_MESSAGES isSuccess:NO];
                }
            }
        }
    }];
}

#pragma mark - PhoneVerificationViewControllerDelegate
- (void)finishedVerifingPhoneWithTarget:(PhoneVerificationViewController *)target {
    //complete signup
    [self tryToSignupUser:target];
}

#pragma mark - DataServiceProtocol
- (void)bind:(id)data forRequestId:(int)rid {
//    // --------------- Legacy actions --------------
}

@end

File: SignUpViewController.swift
//
//  SignUpViewController.swift
//  Bamilo
//
//  Created by Ali Saeedifar on 9/25/18.
//  Copyright © 2018 Rocket Internet. All rights reserved.
//

import UIKit
import Crashlytics

class SignUpViewController: BaseAuthenticationViewCtrl {
    
    private var passwordFieldModel: FormItemModel?
    private var phoneFieldModel: FormItemModel?
    override func viewDidLoad() {
        super.viewDidLoad()
        formParamsName = "customer"
        viewMode = .signUp
        
        tableview.register(UINib(nibName: FormButtonTableViewCell.nibName(), bundle: nil), forCellReuseIdentifier: FormButtonTableViewCell.nibName())
        formController?.submitTitle = STRING_SIGNUP
        
        
        if let email = FormItemModel.email(withFieldName: "\(formParamsName)[email]"),
            let phone = FormItemModel.phone(withFieldName: "\(formParamsName)[phone]"),
            let password =  FormItemModel.passWord(withFieldName: "\(formParamsName)[password]"),
            let nationID =  FormItemModel.nationalID("\(formParamsName)[national_id]") {
            email.placeholder = "\(STRING_EMAIL) *"
            phone.placeholder = "\(STRING_CELLPHONE) *"
            password.placeholder = "\(STRING_PASSWORD) *"
            nationID.placeholder = "\(STRING_NATIONAL_ID) *"
            
            self.passwordFieldModel = password
            
            let signinButtonCell = FormCustomFiled()
            signinButtonCell.cellName = FormButtonTableViewCell.nibName()
            self.formController?.formModelList = [email, phone, password, nationID, "submit", signinButtonCell]
            
            self.phoneFieldModel = phone
        }
        self.formController?.delegate = self
        self.formController?.setupTableView()
    }
    
    
    func requestForSignUp(target: DataServiceProtocol) {
        if var params = self.formController?.getMutableDictionaryOfForm() as? [String: String] {
            AuthenticationDataManager.sharedInstance.signupUser(target, with: &params) { (data, error) in
                if error == nil {
                    self.bind(data, forRequestId: 0)
                    return
                }
                self.errorHandler(error, forRequestID: 0)
                self.trackSignUp(user: nil, success: false)
            }
        }
    }
    
    
    func verifyPhoneForSignUp(with target: DataServiceProtocol, pinCode: String) {
        if let phone = self.phoneFieldModel?.getValue() {
            AuthenticationDataManager.sharedInstance.phoneVerification(target, phone: phone, token: pinCode.convertTo(language: .english)) { (data, errors) in
                if let error = errors {
                    target.errorHandler?(error, forRequestID: 0)
                    return
                }
                self.requestForSignUp(target: target)
            }
        }
    }
    
    private func trackSignUp(user: User?, success: Bool) {
        if let userID = user?.userID {
            Crashlytics.sharedInstance().setUserIdentifier("\(userID)")
        }
        if let name = user?.firstName, let lastName = user?.lastName {
            Crashlytics.sharedInstance().setUserName("\(name) \(lastName)")
        }
        if let email = user?.email {
            Crashlytics.sharedInstance().setUserEmail(email)
        }
        
        TrackerManager.postEvent(selector: EventSelectors.signupEventSelector(), attributes: EventAttributes.signup(method: "normal", user: user, success: success))
    }
}

// MARK: - FormViewControlDelegate
extension SignUpViewController: FormViewControlDelegate {
    func formSubmitButtonTapped() {
        if let valid = self.formController?.isFormValid() {
            if !valid {
                self.formController?.showAnyErrorInForm()
                return
            }
            self.requestForSignUp(target: self)
        }
    }
    
    func customCell(forIndexPath tableView: UITableView!, cellName: String!, indexPath: IndexPath!) -> UITableViewCell! {
        let cell = tableView.dequeueReusableCell(withIdentifier: FormButtonTableViewCell.nibName(), for: indexPath) as! FormButtonTableViewCell
        cell.setTitle(title: STRING_I_HAVE_ACCOUNT)
        cell.tagretMode = .signIn
        cell.delegate = self
        return cell
    }
}

extension SignUpViewController: DataServiceProtocol {
    func bind(_ data: Any!, forRequestId rid: Int32) {
        //TODO: this is not a good approach which has been contracted with server
        // and it's better to be refactored in both platforms
        if let messageList = data as? ApiDataMessageList {
            if let reason = messageList.success?.first?.reason, reason == "CUSTOMER_REGISTRATION_STEP_1_VALIDATED", let phone = phoneFieldModel?.getValue() {
                self.delegate?.requestForVerification(from: .signUp, identifier: phone, target: self, rid: 9)
            }
        }
        //End TODO
        if let password = self.passwordFieldModel?.getValue() {
            if let dictionay = data as? [String: Any], let customerEntity = dictionay[kDataContent] as? CustomerEntity, let customer = customerEntity.entity {
                self.delegate?.successSignUpOrSignInWithUser(user: customer, password: password)
                self.trackSignUp(user: customer, success: true)
            }
            if rid == 0, let dataSource = data as? CustomerEntity, let customer = dataSource.entity {
                self.delegate?.successSignUpOrSignInWithUser(user: customer, password: password)
                self.trackSignUp(user: customer, success: true)
            }
        }
    }
    
    func errorHandler(_ error: Error!, forRequestID rid: Int32) {
        self.baseErrorHandler(error, forRequestID: rid)
    }
    
    func retryAction(_ callBack: RetryHandler!, forRequestId rid: Int32) {
        
    }
}

extension SignUpViewController: FormButtonTableViewCellDelegate {
    func buttonTapped(for target: AuthenticationViewMode) {
        self.delegate?.switchTo(viewMode: target)
    }
}
